client/asset/btc/btc.go

```diff
@@ -1747,336 +1733,206 @@ func (btc *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, time.Time,
-	// If there is no active redemption search for this contract and
-	// the context is canceled while waiting for new blocks to search,
-	// the context cancellation will be caught here and the contract
-	// will be removed from queue to prevent further searches when new
-	// blocks are observed.
-	var result *findRedemptionResult
-	select {
-	case result = <-resultChan:
-	case <-ctx.Done():
-	}
+	outPt := newOutPoint(txHash, vout)
 
-	// If this contract is still in the findRedemptionQueue, remove from the queue
-	// to prevent further redemption search attempts for this contract.
-	btc.findRedemptionMtx.Lock()
-	delete(btc.findRedemptionQueue, contractOutpoint)
-	btc.findRedemptionMtx.Unlock()
+	tx, err := btc.node.getTransaction(txHash)
+	if err != nil {
+		return exitError("getTransaction error for FindRedemption transaction: %v", err)
+	}
 
-	// result would be nil if ctx is canceled or the result channel
-	// is closed without data, which would happen if the redemption
-	// search is aborted when this ExchangeWallet is shut down.
-	if result != nil {
-		return result.RedemptionCoinID, result.Secret, result.Err
+	var blockHash *chainhash.Hash
+	var blockHeight int32 = -1
+	if tx.BlockHash != "" {
+		blockHash, err = chainhash.NewHashFromStr(tx.BlockHash)
+		if err != nil {
+			return exitError("NewHashFromStr error: %v", err)
+		}
+		blockHeight, err = btc.node.getBlockHeight(blockHash)
+		if err != nil {
+			return exitError("GetBlockHeight for redemption %s error: %v", outPt, err)
+		}
 	}
-	return nil, nil, fmt.Errorf("aborted search for redemption of contract %s", contractOutpoint.String())
-}
 
-// queueFindRedemptionRequest extracts the contract hash and tx block (if mined)
-// of the provided contract outpoint, creates a find redemption request and adds
-// it to the findRedemptionQueue. Returns error if a find redemption request is
-// already queued for the contract or if the contract hash or block info cannot
-// be extracted.
-func (btc *ExchangeWallet) queueFindRedemptionRequest(ctx context.Context, contractOutpoint outPoint) (chan *findRedemptionResult, *block, error) {
-	btc.findRedemptionMtx.Lock()
-	defer btc.findRedemptionMtx.Unlock()
+	contractHash := btc.hashContract(contract)
 
-	if _, inQueue := btc.findRedemptionQueue[contractOutpoint]; inQueue {
-		return nil, nil, fmt.Errorf("duplicate find redemption request for %s", contractOutpoint.String())
-	}
-	txHash, vout := contractOutpoint.txHash, contractOutpoint.vout
-	tx, err := btc.node.GetTransaction(txHash.String())
+	contractAddr, err := btc.scriptHashAddress(contract)
 	if err != nil {
-		if isTxNotFoundErr(err) {
-			return nil, nil, asset.CoinNotFoundError
-		}
-		return nil, nil, fmt.Errorf("error finding transaction %s in wallet: %w", txHash, err)
+		return exitError("scriptHashAddress error: %w", err)
 	}
-	msgTx := wire.NewMsgTx(wire.TxVersion)
-	err = msgTx.Deserialize(bytes.NewBuffer(tx.Hex))
+	pkScript, err := txscript.PayToAddrScript(contractAddr)
```

Comment:
This is a little more complicated than that, since we don't actually have the transaction output here. We just have a `GetTransactionResult`, which won't contain any `WalletTxDetails` about the output, since it's to an unimported p2sh address, and also wouldn't contain the pubkey script. As far as I can tell, with a Bitcoin SPV wallet, there's no way to access full transaction data until it's mined, so we can only check the contract once we find the block hash via `getTransaction`. I've added that check in the commit I'm working on now.